### DISCUSSION TITLE
Topology scanner

### DISCUSSION THREAD

**Discussion by agalliazzo** [openthread/openthread#9015] [Show and tell]
Hi Guys,
I don't know if it can be usefull for someone but I just have finished to develop an MVP for a topology monitor.
It is based on python and has a web ui. It uses an openthread device CLI connected to the serial port of the PC to discover the thread topology.

It solved some problem while trying to properly place routers in my house (at the moment routers are stand-alone devices that only do routing, all my thread client node are battery powered)

Please feel free to take a look, download and contribute if you are interested.
https://github.com/agalliazzo/pyThreadTopology

Hope this help someone.
Alessio

**Comment by abtink** [comment]
Thanks @agalliazzo.

Just want to point you to [`otMeshDiag` module](https://github.com/openthread/openthread/blob/main/include/openthread/mesh_diag.h) which can be useful for this. It provides APIs to discover network topology (using the underlying network diag TMF commands). e.g. [`otMeshDiagDiscoverTopology()`](https://github.com/openthread/openthread/blob/d6221f4d326c07f328bb4fc194ef17f1f3d28174/include/openthread/mesh_diag.h#L173-L190). It compiles information about network topology (which device is leader, which devices are BRs, link quality between routers, list of IPv6 addresses of routers, children of each router, etc. as [`otMeshDiagRouterInfo`](https://github.com/openthread/openthread/blob/d6221f4d326c07f328bb4fc194ef17f1f3d28174/include/openthread/mesh_diag.h#L93-L142)).

CLI command [`meshdiag topology` command](https://github.com/openthread/openthread/blob/d6221f4d326c07f328bb4fc194ef17f1f3d28174/src/cli/README.md#meshdiag-topology-ip6-addrschildren) is related to the same API.

There are additional network diagnostic features being added in OT (along with discussion in Thread group to add in Thread 1.3.1 or later version). All related PRs are tracked in 
- https://github.com/openthread/openthread/issues/8460.

**Reply by agalliazzo** [reply]
Thanks for the information @abtink, I was aware about the "new" meshdiag tool, I choose to stay on the safe side due to the eterogenous network I have created.
I mainly have 3 problem:
- My BR installation is more than 1y old, so no big update since then... 
- Some devices are "quite old" too and based on precompiled firmware that I cannot build with the features I want (for example STM32 devices did not implement the DNS resolution and I have to write my own DNS client for the applicaiton processor instead of using the RF processor)
- I'm still approaching to the Thread world, so I took the opportunity to better understand principles that I missed while developing firmware... Let's say that this software, other than an usefull tool for visualizing the topology, is a good way to better understand the whole system


Anyway, for sure, I will implement a legacy scan based on networkdiagnostic and an advanced one based on meshdiag!
Thanks again

**Reply by abtink** [reply]
Sounds good. This tool is, for sure, very useful. Thanks for sharing it. 👍

**Comment by jwhui** [comment]
Very cool, @agalliazzo ! Thanks for sharing.
